icechunk/src/store.rs

```diff
@@ -337,6 +339,57 @@ impl Store {
         }
     }
 
+    pub async fn delete_dir(&self, prefix: &str) -> StoreResult<()> {
+        if self.read_only().await {
+            return Err(StoreError::ReadOnly);
+        }
+
+        // TODO: Handling preceding "/" is ugly!
+        let path = format!("/{}", prefix.trim_start_matches("/"))
+            .try_into()
+            .map_err(|_| StoreError::BadKeyPrefix(prefix.to_owned()))?;
+        let res = {
+            let guard = self.session.read().await;
+            let node = guard.get_node(&path).await;
+            match node {
+                Ok(NodeSnapshot {
+                    node_data: NodeData::Group, path: node_path, ..
+                }) => Ok((NodeType::Group, node_path)),
+                Ok(NodeSnapshot {
+                    node_data: NodeData::Array(..),
+                    path: node_path,
+                    ..
+                }) => Ok((NodeType::Array, node_path)),
+                Err(err) => Err(err),
+            }
+        };
+
+        match res {
+            Ok((NodeType::Group, node_path)) => {
+                let mut guard = self.session.write().await;
+                Ok(guard.deref_mut().delete_group(node_path).await?)
+            }
+            Ok((NodeType::Array, node_path)) => {
+                let mut guard = self.session.write().await;
+                Ok(guard.deref_mut().delete_array(node_path).await?)
+            }
+            _ => {
+                // other cases are
+                // 1. delete("/path/to/array/c")
+                // 2. delete("/path/to/array/c/0/0")
+                let to_delete = self
+                    .list_prefix(prefix.trim_start_matches("/"))
+                    .await?
+                    .try_collect::<Vec<_>>()
+                    .await?;
+                for item in to_delete {
+                    self.delete(&item).await?;
```

Comment:
Are these guaranteed to exist when delete is called? Or should we just catch the error in case they are deleted out of order